atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/TokenRangeFetcher.java

```diff
@@ -0,0 +1,97 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.RangeSet;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
+import com.palantir.atlasdb.cassandra.CassandraServersConfigs;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraConstants;
+import com.palantir.atlasdb.keyvalue.cassandra.LightweightOppToken;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+final class TokenRangeFetcher {
+    private static final SafeLogger log = SafeLoggerFactory.get(TokenRangeFetcher.class);
+
+    private static final int LONG_READ_TIMEOUT_MS = (int) TimeUnit.MINUTES.toMillis(2);
+    // reduce this from default because we run RepairTableTask across N keyspaces at the same time
+    private static final int SELECT_FETCH_SIZE = 1_000;
+
+    private final CqlSession cqlSession;
+    private final CqlMetadata cqlMetadata;
+    private final CassandraKeyValueServiceConfig config;
+
+    public TokenRangeFetcher(CqlSession cqlSession, CassandraKeyValueServiceConfig config) {
+        this.config = config;
+        this.cqlSession = cqlSession;
+        this.cqlMetadata = cqlSession.getMetadata();
+    }
+
+    public Map<InetSocketAddress, RangeSet<LightweightOppToken>> getTokenRange(String tableName) {
+        String keyspaceName = config.getKeyspaceOrThrow();
+        KeyspaceMetadata keyspace = cqlMetadata.getKeyspaceMetadata(keyspaceName);
+        TableMetadata tableMetadata = keyspace.getTable(tableName);
+        Set<LightweightOppToken> partitionTokens = getPartitionTokens(tableMetadata);
+        Map<InetSocketAddress, RangeSet<LightweightOppToken>> tokenRangesByNode = ClusterMetadataUtils.getTokenMapping(
+                CassandraServersConfigs.getCqlHosts(config), cqlMetadata, keyspaceName, partitionTokens);
+
+        if (!partitionTokens.isEmpty() && log.isDebugEnabled()) {
+            int numTokenRanges = tokenRangesByNode.values().stream()
+                    .mapToInt(ranges -> ranges.asRanges().size())
+                    .sum();
+
+            log.debug(
+                    "Identified token ranges requiring repair",
+                    SafeArg.of("keyspace", keyspace),
+                    SafeArg.of("table", tableName),
+                    SafeArg.of("numPartitionKeys", partitionTokens.size()),
+                    SafeArg.of("numTokenRanges", numTokenRanges));
+        }
+
+        return tokenRangesByNode;
+    }
+
+    private Set<LightweightOppToken> getPartitionTokens(TableMetadata tableMetadata) {
+        return cqlSession.retrieveRowKeysAtConsistencyAll(createSelectStatements(tableMetadata));
+    }
+
+    private static List<Statement> createSelectStatements(TableMetadata table) {
```

Comment:
general cleanup: I realise this is copied from the original, but yeah let's say what the statements do